### DISCUSSION TITLE
Jib vs. Cloud Native Buildpacks

### DISCUSSION THREAD

**Discussion by timothystone-knsl** [jhipster/generator-jhipster#26527] [Ideas]
I stumbled, figuratively, on the availability of Cloud Native Buildpacks (CNB) in Spring Boot 2.3+ today. `spring-boot:build-image` and `spring-boot:build-image-no-fork` both support creating a container from the project and are available "natively" to Spring Boot without an extra plugin, currently the jib-maven-plugin.

I'm a fan of Jib and equally a supporter of JHipster's use of Jib. Heck, I wrote a CI/CD component for Jib and contributed it to [To Be Continuous's Maven template](https://to-be-continuous.gitlab.io/doc/ref/maven/#jib-variant). I'm also not aware if the Jib support in JHipster predates Spring Boot's CNB support.

To achieve parity looks doable. Removing Jib could be a blow to the visibility of Jib (I know JHipster made me aware of it), but it could simplify the stack by using one less tool and track to Spring Boot's built-in support as well. I may take an inventory of existing documentation and generator templates to judge the LOE and open a PR.

Wholesale replacement or generator option would be the next question. Currently, one get's Jib without even being aware of it, so replacement might be desirable. 

Thoughts?

**Comment by mraible** [comment]
I'm fine with making the change. I'm not sure how much documentation, workflows, and code needs to be updated.

**Comment by timothystone-knsl** [comment]
Research so far: Jib was introduced to JHipster prior to 2019. The [earliest commit](4372171992dc706b42a3631c4057e197603ddd5a) of `jib-maven-plugin` I found in GitHub was an update to 1.7.0 of the plugin by @DanielFran. Spring introduced [CNB in version v2.3.0.RELEASE](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.3-Release-Notes#build-oci-images-with-cloud-native-buildpacks) c. 2020. This supports my prior guess that Jib provides tooling that was not available in Spring Boot when it was introduced.

My initial examination of `spring-boot:build-image(-no-fork)?` shows that there are some interesting default choices in the Paketo Buildpacks like the JDK/JRE vendor and the base image of Ubuntu. This all looks very configurable of course, but just coming to it needs some ramp up time. @atomfrede said, in another discussion, that JHipster's default position is "defaults over opinions" which I would honor (but "bellsoft-liberica," never heard of them as a JDK/JRE vendor; Adoptium, Azul, et al., yes. I'm sure it's great, but not a first in mind offering for a "default").

Jib really leans into my prejudices: Jib's POM declarative configuration feels more natural to me (as a Maven early adopter).

This idea may simmer a bit longer. I might start working on a PR in my fork, but may not submit anything in the immediate future. Anyone that's passionate about CNB on the core team is welcome to "steal this idea" and run with it.

**Reply by atomfrede** [reply]
Yes, I would definitly try the cloud native build packs. I think for the native blueprint we use it already, so going with the default spring offers makes sense with [regards to our jhipster policy](https://www.jhipster.tech/policies/#policy-1-technologies-used-by-jhipster-have-their-default-configuration-and-best-practices-used-as-much-as-possible) .
And totally right, we never revisited the jib usage after it has been added for a good reason (no default solution by the spring team available).

**Reply by atomfrede** [reply]
Maybe I will give it a try, as that would also simplify our build file configurations

**Reply by timothystone-knsl** [reply]
Please share any repo or branch with me. Happy to contribute! I could learn a few things along the way I'm sure about JHipster development. I also contribute as @timothystone. "-knsl" is my company ID and usually appears when I'm working on my employer laptop.

**Reply by DanielFran** [reply]
JIB was integrated in 2018 (see https://github.com/jhipster/generator-jhipster/pull/8352) by @danielpetisme because the previous gradle-docker-plugin had CVE issues reported (see https://github.com/jhipster/generator-jhipster/issues/8248).
At that time, it was the only alternative and could be used on both Gradle and Maven.
It was a good option since [dockerfile-maven-plugin](https://github.com/spotify/dockerfile-maven) was also deprecated a few months later.

**Reply by timothystone-knsl** [reply]
Lest it be lost, I love the Jib integration. See OP:

> I'm a fan of Jib and equally a supporter of JHipster's use of Jib. Heck, I wrote a CI/CD component for Jib and contributed it to [To Be Continuous's Maven template](https://to-be-continuous.gitlab.io/doc/ref/maven/#jib-variant). I'm also not aware if the Jib support in JHipster predates Spring Boot's CNB support.

Question asked and answered. Thank @DanielFran.

I saw the availability of CNB via Spring Boot and thought 🤔 💭  "Hmmm..." and wanted to start a conversation. To Be Continuous even supports CNB so my org's CI/CD would not change much.

**Comment by atomfrede** [comment]
I came just across this post https://twitter.com/helpermethod/status/1808857739092725840. If CDS does not work (easily) with jib but with buildpacks, that would be a huge argument in favor to move to boot's default buildpack support. https://docs.spring.io/spring-framework/reference/integration/cds.html
